### PR TITLE
feat: hint options for gRPC insert

### LINKED ISSUE #4441: gRPC supports create table hints
### What type of enhancement is this?

API improvement

### What does the enhancement do?

When ingesting data by gRPC ( [Java ingester](https://github.com/GreptimeTeam/greptimedb-ingester-java)  or [Erlang Ingestor](https://github.com/GreptimeTeam/greptimedb-ingester-erl)), the tables will be automatically created by the data schema.

But in some scenarios, we want to set table options such as `append_mode='true'` or `merge_mode='last_non_null'` like that, it's better to support the hints or flags for these important options which will affect the behavior or performance.

### Implementation challenges

_No response_

### PR DESCRIPTION
**Pull request by fengjiachun**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

This PR closes #4441 

# How to use:
For example, if you need to add a hint during the first write to automatically create a table with the option `merge_mode = "last_non_null"`, then we need to add a hint to the gRPC request header:

```
("x-greptime-hint-merge_mode", "last_non_null")
```

Note, you must use the prefix `x-greptime-hint-`; otherwise, it will not be parsed as a valid hint.

Taking the java ingester client as an example, we modify a line of code in this example to:

https://github.com/GreptimeTeam/greptimedb-ingester-java/blob/4fa729a537df3f49278aef58fa919315ad6a7f88/ingester-example/src/main/java/io/greptime/LowLevelApiWriteQuickStart.java#L76

```
CompletableFuture<Result<WriteOk, Err>> future = greptimeDB.write(cpuMetric, memMetric);
```
->
```
Context ctx = Context.hint("merge_mode", "last_non_null");
CompletableFuture<Result<WriteOk, Err>> future = greptimeDB.write(Arrays.asList(cpuMetric, memMetric), WriteOp.Insert, ctx);
```

Currently, only two hint options are supported, which are append_mode and merge_mode.

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [x] This PR requires documentation updates.

### CODE REVIEW

**Comment by killme2008** [comment]
Good job! A little Java API suggestion:

```java
Context.of("x-greptime-hint:merge_mode", "last_non_null"));
```

Adds an API specified for hints:

```
Context.hint("merge_mode", "last_non_null"));
```

File: src/servers/src/grpc/database.rs [hunk 1]
```diff
@@ -33,6 +36,31 @@ impl DatabaseService {
...
+                let Ok(value) = value.to_str() else {
+                    // Simply return None for non-string values.
+                    return None;
+                };
+                // Safety: we already checked the prefix.
+                let new_key = key
+                    .strip_prefix(GREPTIME_DB_HEADER_HINT_PREFIX)
+                    .unwrap()
+                    .to_string();
+                Some((new_key, value.to_string()))
```
Comment (killme2008) [inline src/servers/src/grpc/database.rs:60]: Better to trim the value.

Comment (killme2008) [reply src/servers/src/grpc/database.rs:60]: Adds a unit test for this function.

**Review by killme2008 — COMMENTED** [review]
I think we can add a test to ensure the API working well.

https://github.com/GreptimeTeam/greptimedb/blob/main/tests-integration/tests/http.rs

**Comment by fengjiachun** [comment]
> Good job! A little Java API suggestion:
> 
> ```java
> Context.of("x-greptime-hint:merge_mode", "last_non_null"));
> ```
> 
> Adds an API specified for hints:
> 
> ```
> Context.hint("merge_mode", "last_non_null"));
> ```

Good point!

**Review by killme2008 — APPROVED** [review]
LGTM

**Comment by killme2008** [comment]
PTAL @WenyXu

**Review by WenyXu — APPROVED** [review]
LGTM

File: src/servers/src/grpc/database.rs [hunk 2]
```diff
@@ -129,3 +142,58 @@ impl GreptimeDatabase for DatabaseService {
...
+            }
+            let Ok(value) = value.to_str() else {
+                // Simply return None for non-string values.
+                return None;
+            };
+            // Safety: we already checked the prefix.
+            let new_key = key
+                .strip_prefix(GREPTIME_DB_HEADER_HINT_PREFIX)
+                .unwrap()
+                .to_string();
```
[the remark this answers is not in the document]
Comment (fengjiachun) [reply src/servers/src/grpc/database.rs:165]: Looks good, I'll do it in next PR

**Pull request merged**